docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+stale? How do you detect conflicts between two relocation operations initiated
+from different places?
+
+This RFC suggests that no such relocation state should be persisted. Instead, a
+system that wants to initiate a relocation will perform only the first
+operation, and a different system will detect the need for a complementary
+operation and perform it. A relocation is thus completed without any state being
+exchanged between those two systems.
+
+By eliminating the need to persist any data about in-progress relocation
+operations, the overall system is dramatically simplified.
+
+# Detailed design
+The implementation involves a few pieces:
+
+1. Each range must have a persisted target replication state. This does not
+   prescribe specific replica locations; it suggests only a desired number of
+   replicas and some desired attributes of the stores they are placed on.
+2. The core mechanic is a stateless function which can compare the immediate
+   replication state of a range to its target replication state; if the desired
+   state is different, this function will either create or remove replicas until
+   the target state is matched.
+3. Finally, any operations that wish to relocate a replica need only perform the
+   first operation of the relocation (either a create or a remove). This will
+   perturb the range's replication state away from its target; the core function
+   will later detect the mismatch, and correct it by performing the
+   complementary operation of the relocation (a remove or a create).
+
+The first piece is already present: each range has a zone configuration
+which determines the target replication state for the range.
+
+The second piece, the core mechanic, will be performed by the replicate
+queue. This queue is already used to add replicas to ranges which are
+under-replicated; it can be enhanced to remove replicas from over-replicated
+ranges, thus satisfying the basic requirements of the core mechanic. Detailed
+design for these changes are in a [separate RFC (#2153)](https://github.com/cockroachdb/cockroach/pull/2153).
+
+The third piece simply informs the design of systems performing relocations:
+most immediately, these will be the rebalance system and the repair system.
+These systems, when identifying a relocation opportunity, will simply perform
+the first operation of the relocation, and the replicate queue will perform the
+complementary operation. 
+
+RFCs for these operations are forthcoming.
+
+The final complication is how to ensure that the replicate queue promptly
+identifies ranges outside of their ideal target state. As a queue it will be
+attached to the replica scanner, but we will also want to enqueue a replica
+immediately whenever we knowingly perturb the replication state. Thus,
+components initating a relocation (e.g. rebalance or repair) should immediately
+enqueue their target replicas after changing the replication state. 
```

Comment:
I'd say this is already known. A halfway-completed replication operation leaves the range in an unstable state due to quorum requirements; we should aim to resolve this state in seconds, not minutes.